File: solutions/fsharp/yacht/1/Yacht.fs
module Yacht

type Category = 
    | Ones
    | Twos
    | Threes
    | Fours
    | Fives
    | Sixes
    | FullHouse
    | FourOfAKind
    | LittleStraight
    | BigStraight
    | Choice
    | Yacht

type Die =
    | One 
    | Two 
    | Three
    | Four 
    | Five 
    | Six

let score category dice =

    let dieValue die =
        match die with
        | Die.One -> 1
        | Die.Two -> 2
        | Die.Three -> 3
        | Die.Four -> 4
        | Die.Five -> 5
        | Die.Six -> 6

    let sumValues die dice =
        dice
        |> Seq.filter (fun d -> d = die)
        |> Seq.sumBy dieValue

    let sortedDice dice = 
        dice
        |> Seq.sortBy (fun v -> dieValue v)
        |> Seq.toList

    let groupCount dice =
        dice
        |> Seq.groupBy (fun d -> d)
        |> Seq.map (fun (die, lst) -> (die, Seq.length lst))

    let (|Has4OrMore|_|) dice = 
        dice
        |> groupCount
        |> Seq.filter (fun (_, cnt) -> cnt >= 4)
        |> Seq.tryExactlyOne

    let HasFullHouse dice =
        let counts =
            dice
            |> groupCount
            |> Seq.map (fun (_, n) -> n)
            |> Seq.sort
            |> Seq.toList

        counts = [2; 3]

    match (category, dice) with
    | (Category.Yacht, _) when Seq.length (Seq.distinct dice) = 1 -> 50
    | (Category.LittleStraight, _) when sortedDice dice = [Die.One; Die.Two; Die.Three; Die.Four; Die.Five] -> 30
    | (Category.BigStraight, _) when sortedDice dice = [Die.Two; Die.Three; Die.Four; Die.Five; Die.Six] -> 30
    | (Category.Choice, _) -> Seq.sumBy dieValue dice
    | (Category.Ones, _) -> sumValues Die.One dice
    | (Category.Twos, _) -> sumValues Die.Two dice
    | (Category.Threes, _) -> sumValues Die.Three dice
    | (Category.Fours, _) -> sumValues Die.Four dice
    | (Category.Fives, _) -> sumValues Die.Five dice
    | (Category.Sixes, _) -> sumValues Die.Six dice
    | (Category.FourOfAKind, Has4OrMore (die, _)) -> (dieValue die) * 4
    | (Category.FullHouse, _) when HasFullHouse dice -> Seq.sumBy dieValue dice
    | _ -> 0

File: solutions/csharp/yacht/1/Yacht.cs
using System;
using System.Linq;

public enum YachtCategory
{
    Ones = 1,
    Twos = 2,
    Threes = 3,
    Fours = 4,
    Fives = 5,
    Sixes = 6,
    FullHouse = 7,
    FourOfAKind = 8,
    LittleStraight = 9,
    BigStraight = 10,
    Choice = 11,
    Yacht = 12,
}

public static class YachtGame
{
    public static int Score(int[] dice, YachtCategory category)
        => category switch {
            YachtCategory.Yacht when IsYacht(dice) => 50,
            YachtCategory.Ones => CountSingle(dice, 1),
            YachtCategory.Twos => CountSingle(dice, 2),
            YachtCategory.Threes => CountSingle(dice, 3),
            YachtCategory.Fours => CountSingle(dice, 4),
            YachtCategory.Fives => CountSingle(dice, 5),
            YachtCategory.Sixes => CountSingle(dice, 6),
            YachtCategory.FullHouse => CountFullHouse(dice),
            YachtCategory.FourOfAKind => Count4OrMore(dice),
            YachtCategory.LittleStraight when IsLittleStraight(dice) => 30,
            YachtCategory.BigStraight when IsBigStraight(dice) => 30,
            YachtCategory.Choice => dice.Sum(),
            _ => 0,
        };

    private static bool IsYacht(int[] dice)
        => dice.Distinct().Count() == 1;

    private static int CountSingle(int[] dice, int value)
        => dice.Where(d => d == value).Sum();

    private static int CountFullHouse(int[] dice)
    {
        var grouped = dice
            .GroupBy(g => g)
            .Select(g => g.Count() )
            .OrderDescending()
            .ToList();

        if (grouped.SequenceEqual(new[] {3, 2}))
            return dice.Sum();

        return 0;
    }

    private static int Count4OrMore(int[] dice)
    {
        var selected = dice
            .GroupBy(g => g)
            .Select(g => new { count = g.Count(), score = g.First() * 4 } )
            .SingleOrDefault(s => s.count >= 4);

        return selected?.score ?? 0;
    }

    private static bool IsLittleStraight(int[] dice)
        =>
            dice
                .Order()
                .SequenceEqual(new[] {1, 2, 3, 4, 5});

    private static bool IsBigStraight(int[] dice)
        =>
            dice
                .Order()
                .SequenceEqual(new[] {2, 3, 4, 5, 6});
}

